hook

Sasasu opened the pull request:
## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

Sasasu committed:
load libaray in both bootstrap and single mode

Signed-off-by: Sasasu <i@sasa.su>

Sasasu committed:
gpinit: load extension from environment

Signed-off-by: Sasasu <i@sasa.su>

Sasasu committed:
ao: add hooks for data encryption

Signed-off-by: Sasasu <i@sasa.su>

Sasasu committed:
heap: add hooks for data encryption

Signed-off-by: Sasasu <i@sasa.su>

Commit:
Remove unused buf[MAXPGPATH] variable in mdcreate_ao()

beeender commented on gpcontrib/Makefile:
```diff
@@ -22,7 +22,9 @@ ifeq "$(enable_debug_extensions)" "yes"
                gp_toolkit \
-               pg_hint_plan
+               pg_hint_plan \
+               gp_toolkit \
```
two `gp_toolkit`, the other one at line 24

dgkimura committed:
[ORCA] Fix ident to const predicate push down optimization

In preprocessing, ORCA may choose to add select filters in order to help
trigger predicate push down during query normalization. In preprocessing
step PexprReplaceColWithConst(), a duplicate filter is created of join
predicates, with idents replaced by consts from equality predicates
among the select filter. The transformation looks like:

  Input:
    Select
      |-- NaryJoin/LeftOuterJoin
      |   |-- ...
      |   +-- Join Pred (has idents)
      +-- Filter

  Output:
    Select
      |-- Select
      |  |-- NaryJoin/LeftOuterJoin
      |  |  |-- ...
      |  |  +-- Join Pred (has idents)
      |  +-- Join Pred (has consts)
      +-- Filter

However, an issue occurs if the Join Pred contains a CTEAnchor. In that
case the CTEAnchor is "duplicated" and leads to an invalid plan. In
retail build that manifests as a crash during DXL to PlStmt translation.
This commit fixes the issue by choosing to not add the select filter
when the Join Pred contains a CTEAnchor.

chrishajas committed:
Re-support DISTINCT-qualified Window Aggregate in Orca (#17113)

This was enabled in 5X, but was lost in the 6X merge and Orca fell back to Planner. This was re-enabled in the executor in 6523b43, but was not added for Orca. 386d542 enables the fallback in Orca.

In 5X, Orca and planner would generate the following plan for the query
create table t2 (a int, b int, c int) distributed by (c);
explain select a, count( distinct b) over (partition by a) from t2;

```
 Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1293.01 rows=20 width=12)
   ->  Sequence  (cost=0.00..1293.01 rows=7 width=12)
         ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=7 width=1)
               ->  Materialize  (cost=0.00..431.00 rows=7 width=1)
                     ->  Redistribute Motion 3:3  (slice1; segments: 3)  (cost=0.00..431.00 rows=7 width=42)
                           Hash Key: t2.a
                           ->  Table Scan on t2  (cost=0.00..431.00 rows=7 width=42)
         ->  Hash Join  (cost=0.00..862.00 rows=7 width=12)
               Hash Cond: NOT share0_ref3.a IS DISTINCT FROM share0_ref2.a
               ->  GroupAggregate  (cost=0.00..431.00 rows=4 width=12)
                     Group By: share0_ref3.a
                     ->  Sort  (cost=0.00..431.00 rows=7 width=8)
                           Sort Key: share0_ref3.a
                           ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=7 width=8)
               ->  Hash  (cost=431.00..431.00 rows=7 width=4)
                     ->  Shared Scan (share slice:id 2:0)  (cost=0.00..431.00 rows=7 width=4)
 Optimizer status: PQO version 3.117.0
(17 rows)
```

Now, we'll generate the following plan in both Orca and Planner:
```
test=# explain select a, count( distinct b) over (partition by a) from t2;
                                              QUERY PLAN
------------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..431.00 rows=1 width=12)
   ->  WindowAgg  (cost=0.00..431.00 rows=1 width=12)
         Partition By: a
         ->  Sort  (cost=0.00..431.00 rows=1 width=8)
               Sort Key: a
               ->  Redistribute Motion 3:3  (slice2; segments: 3)  (cost=0.00..431.00 rows=1 width=8)
                     Hash Key: a
                     ->  Seq Scan on t2  (cost=0.00..431.00 rows=1 width=8)
 Optimizer: GPORCA
(9 rows)
```
The windistinct flag is passed through the WindowAgg node, and the distinct operation is handled within the executor.
This was already done in the Orca code, and the logic was already implemented. However, we still had the preprocessing logic. We simply needed to stop doing this extra logic within the optimizer (in the preprocessor) since it's now done in the executor.

Basic performance tests show that this implementation is also faster than the 5X plan for both Orca and Planner when there is a PARTITION BY clause for the window function

24nishant committed:
Prevent overwriting of workers in ORCA (#17097)

In some cases, during optimization of queries with call to a function, it is observed that another instance of ORCA optimization is launched.

At the start of optimization, a new worker is registered in the 'Worker Pool Manager'. Currently ORCA supports only one worker, if a new worker is registered on top of an existing worker, then the details of the old worker is lost, causing undefined behaviour. Thus, through this PR, aim is to stop overwriting of workers.

---------

Co-authored-by: nishant sharma <nishants4@nishants46M6RY.vmware.com>

Sasasu committed:
gpcontrib: add data_encryption as example

Signed-off-by: Sasasu <i@sasa.su>
